README.md
# authnet-clj - A native Clojure Authorize.Net client

Authorize.Net Clojure - AuthNet-Clj - is a native Clojure client for the Authorize.Net JSON API. It aims to cover mostly the ARB, CIM, and Transactions end-points.

|Version|1.0.1|
|----|-----|
|Docs|https://erniedeferia.github.io/authnet-clj/|
|Source|https://github.com/erniedeferia/authnet-clj|


## Usage

```clojure
[authnet-clj "0.1.0"]

(ns my.app.namespace
  (:require [authnet-clj :refer :all]))


;; example 1: to get suspended subscriptions

(get-suspended-subscriptions)

;; exmaple 2: to update payment method (credit card) on existing subscription

(update-subscription "123456" "370000000000002" "2020-12")

```

## Configuration

AuthNet-Clj expects the following Authorize.Net credentials as environment variables.

```sh
AUTHORIZENET_GATEWAY=sandbox
AUTHORIZENET_LOGIN_ID
AUTHORIZENET_TRANSATION_KEY
```

When `AUTHORIZENET_GATEWAY` is set to sandbox, `AutNet-Clj` will operate on the test (or sandbox)
environment. A sandbox account can be created at the following location:


https://developer.authorize.net/hello_world/sandbox/



# Authorize.Net API Documentation & Reference

https://developer.authorize.net/api/reference
https://developer.authorize.net/api/reference/index.html#recurring-billing-update-a-subscription


Copyright © 2020 MACROFEX LLC



CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- NA

### Removed
- N/A

### Fixed
- N/A

## 0.1.0 - 2020-01-30
### Added
- Initial version with support for querying the ARB (Automated Recurring Billing) API
- Partial support for the Customer Information Management (CIM) interface

[Unreleased]: https://github.com/your-name/subsmgr/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/subsmgr/compare/0.1.0...0.1.1


test/authnet_clj/core_test.clj
(ns authnet-clj.core-test
  (:require [clojure.test :refer :all]
            [authnet-clj.core :refer :all]))

(deftest a-test
  (testing "FIXME, I fail."
    (is (= 0 1))))


src/authnet_clj/arb.clj
;; ## ARB - Automatic Recurring Billing
;;
;; This namespace implements the functions require to list, get, create, and update
;; recurring billing subscriptions.
;;
(ns authnet-clj.arb
  (:require [authnet-clj.core :refer :all] )
  )


;; # Records to facilitate subscription update.
;;
;; The following example illustrates the basic structure for
;; a subscription payment update. The records defined below
;; reflect each of the segments of this structure.
;;
;; <pre>
;; <code>
;; {
;;     "ARBUpdateSubscriptionRequest": {
;;         "merchantAuthentication": {
;;             "name": "********",
;;             "transactionKey": "*******************"
;;         },
;;         "refId": "123456",
;;         "subscriptionId": "100748",
;;         "subscription": {
;;             "payment": {
;;                 "creditCard": {
;;                     "cardNumber": "4111111111111111",
;;                     "expirationDate": "2020-12"
;;                 }
;;             }
;;         }
;;     }
;; }
;; </code>
;; </pre>


;; The billing address associated with a credit card update
(defrecord billTo [firstName lastName address city state zip])

;; Credit card record with card number and expiration date attributes
(defrecord creditCard [cardNumber expirationDate])

;; A payment segnment includes a credit card map
(defrecord payment [^creditCard creditCard])

;; A subscription update can include a new CC as well as
;; a new billTo (name, address). The `update-subscription`
;; function has two arities to support these two scenarios.
(defrecord subscription [^payment payment ^billTo billTo])


(defn make-subscription-payment-update
  " Makes the map required to submit a subscription payment method update request.
  "
 ([credit-card expiration-date]
  (->subscription (->payment (->creditCard credit-card expiration-date))))
 ([credit-card expiration-date f-name l-name address city state zip]
  (->subscription (->payment (->creditCard credit-card expiration-date)) (->billTo f-name l-name address city state zip)))
  )


;;
;; ## Concrete Subscription Request Functions
;; <br/>

(defn get-subscription-status
  "Retrieves the subscription status for the specified subscription.
  "
  [subscription-id]
  (let [creds (get-credentials)
        url (get-endpoint creds)]
    (do-request url (build-request :ARBGetSubscriptionStatusRequest {:subscriptionId subscription-id} ) )
    )
  )

(defn get-subscription
  "Retrieves the details of the specified subscription.
  "
  [subscription-id]
  (let [creds (get-credentials)
        url (get-endpoint creds)]
    (do-request url (build-request :ARBGetSubscriptionRequest {:subscriptionId subscription-id} ) )
    )
  )


(defn get-subscriptions
  " Retrieves a sequence of subscriptions in the specified status.

  Inputs:
     <pre>
     search-type: cardExpiringThisMonth
                  subscriptionActive
                  subscriptionInactive
                  subscriptionExpiringThisMonth
     </pre>
  "
  [search-type]
  (let [creds (get-credentials)
        url (get-endpoint creds)]
    (do-request url (build-request :ARBGetSubscriptionListRequest {:searchType search-type }))
    )
  )


(defn update-subscription
  " Updates the Credit Card number and Expirate Date of the specified subscription.
  It can also update the Credit Card holder's name and address if 9-arity function
  is invoked.
  "
  ([subscription-id cc-number cc-expiration]
   (let [creds (get-credentials)
         url (get-endpoint creds)
         ]
     (do-request url (build-request :ARBUpdateSubscriptionRequest
                                    {:subscriptionId subscription-id
                                     :subscription (make-subscription-payment-update cc-number cc-expiration )}))))
  ([subscription-id cc-number cc-expiration fname lname address city state zip]
   (let [creds (get-credentials)
         url (get-endpoint creds)
         ]
     (do-request url (build-request :ARBUpdateSubscriptionRequest
                                    {:subscriptionId subscription-id
                                     :subscription (make-subscription-payment-update cc-number cc-expiration fname lname address city state zip)})))
   ))

(defn cancel-subscription
  "Makes a request to cancel the identified subscription.
  "
  [subscription-id]
  (let [creds (get-credentials)
        url (get-endpoint creds)
        ]
    (do-request url  (build-request :ARBCancelSubscriptionRequest {:subscriptionId subscription-id}))
    )
  )


src/authnet_clj/profiles.clj
;; ## Customer Profile
;;
;; This namespace implements the functions related to retrieving and updating
;; customer profiles.
;;
(ns authnet-clj.profiles
  (:require [authnet-clj.core :refer :all])
  )


(defn get-customer-profile
  "Retrieves a customer profile."
  [customer-profile-id]
  (let [creds (get-credentials)
        url (get-endpoint creds)
        ]
    (do-request url (build-request :getCustomerProfileRequest {:customerProfileId customer-profile-id :includeIssuerInfo "true"}))
    )
  )

(defn list-expiring-customers-payment-profiles
  "Lists all payment profiles that expire in the specified month.
     month: \"2020-12\"
     </pre>
  "
  [month]
  (let [creds (get-credentials)
        url (get-endpoint creds)
        ]
    (do-request url (build-request :getCustomerPaymentProfileListRequest {:searchType "cardsExpiringInMonth" :month month}))
    ))

(defn list-customers-profile-ids
  "Lists all customers profiles IDs."
  []
  (let [creds (get-credentials)
        url (get-endpoint creds)
        ]
    (do-request url (build-request :getCustomerProfileIdsRequest {} ))
    )
  )


src/authnet_clj/core.clj
;; ## Core Common functions
;;
;; This namespace holds core functions used by each of the (ARB, CIM, Trans)
;; request-specific namespaces.
;;
(ns authnet-clj.core
  (:require [cheshire.core :refer :all ]
            [clj-http.client :as client]
            )
  (:gen-class))


(def endpoints
  "Defines the sandbox and production end-points"
  { :sandbox "https://apitest.authorize.net/xml/v1/request.api"
    :production "https://api.authorize.net/xml/v1/request.api" })

(defn get-credentials
  "Retrieves the Authorize.net credentials from environment variables."
  []
  (let [name (System/getenv "AUTHORIZENET_LOGIN_ID")
        key (System/getenv "AUTHORIZENET_TRANSACTION_KEY")
        gateway (System/getenv "AUTHORIZENET_GATEWAY")
        ]

    (if (nil? name)
      (throw (ex-info "Environment variable AUTHORIZENET_LOGIN_ID is missing" {}))
      )

    (if (nil? key)
      (throw (ex-info "Environment variable AUTHORIZENET_TRANSACTION_KEY is missing" {}))
      )

    (if (nil? gateway)
      (throw (ex-info "Environment variable AUTHORIZENET_GATEWAY is missing" {}))
      )

  {:name name
   :transactionKey key
   :gateway gateway } )
)

;; Helper functions to get each Authorize.net credential variable.
(defn get-name [creds] (:name (get-credentials)))
(defn get-transaction-key [creds] (:transactionKey (get-credentials)))
(defn get-gateway [creds] (keyword (:gateway (get-credentials)) )  )
(defn get-endpoint [creds]
  ((get-gateway creds) endpoints))


(defn get-marchant-authentication-map
  "Common transaction request maps generators. These are used
  within each of the API segements (ARB, Customer Profile, etc.)
  "
  [creds]
  {:merchantAuthentication
   {:name (:name creds) :transactionKey (:transactionKey creds)}}
  )

(defn get-transaction-map
  "Generates the complete transaction request map including the transaction name along
  with the merchant Authentication segment followed by the transaction payload.

  Inputs:
  -------
  transaction-name: a keyword, such as :ARBGetSubscriptionstatusrequest that identifies
                    the transactions to be submitted.

  merchantauth-map: a map with a single key :merchantAuthentication and which value
                    is a map containing the API Login ID and API Transaction Key.

  Note: Use the get-merchant-authentication-map helper to generate the merchantauth-map.

  payload-map:      a map of k/v that is specific to the transaction (transaction-name).
                    (https://developer.authorize.net/api/reference)

  Output:
  -------
  A map similar to the one depicted below, but with the k/v that are specific
  to a transaction after the merchantAuthentication map.

  <pre>
  <code>
  {
    'ARBGetSubscriptionStatusRequest': {
        'merchantAuthentication': {
            'name': '<AUTHORIZINET_LOGIN_ID>',
            'transactionKey': '<AUTHORIZENET_TRANSACTION_KEY>'
        },
        'refId': '123456',
        'subscriptionId': '100748'
    }
  }
  </code>
  </pre>

  "
  [merchantauth-map transaction-name payload-map]
  {transaction-name  (merge  merchantauth-map payload-map) }
  )

(defn build-request
  "Request builder responsible for creating a properly formed
  request map including the merchant Authentication segment.
  "
  [request-type request-payload]
  (-> (get-credentials)
      (get-marchant-authentication-map)
      (get-transaction-map request-type request-payload )
      )
  )

(defn do-request
 "Request producer responsible for issuing the POST request to the Authorize.Net end-point
  specified via the environment variables.

  Note: Authorize.Net returns BOM (Byte Order Marker) byte that screws
  up the JSON parser. The combination of reading the body as
  a stream and removing the BOM \uFEFF does the trick.
  "
  [url query]
  (let [res (client/post url
               {:body (generate-string query )
                :accept :json
                :body-encoding "UTF-16"
                :as :stream
                :throw-exceptions? false
                :content-type :json}) ]
    (-> res
        (:body)
        (slurp)
        (clojure.string/replace #"\r\n|\n|\r|\uFEFF" "")
        (parse-string true)
        )
    )
  )
